runtimeconfig/manager.go

```diff
@@ -118,13 +130,108 @@ func (om *Manager) CloseListenerChannel(listener <-chan interface{}) {
 	}
 }
 
-func (om *Manager) loop(ctx context.Context) error {
+func (om *Manager) detectChanges(ctx context.Context) error {
 	if om.cfg.LoadPath == "" {
 		level.Info(om.logger).Log("msg", "runtime config disabled: file not specified")
 		<-ctx.Done()
 		return nil
 	}
 
+	if om.cfg.ChangeDetector == "" || om.cfg.ChangeDetector == Poll {
+		om.pollLoop(ctx, func(error) bool { return false })
+	} else if om.cfg.ChangeDetector == Notify {
+		err := om.notifyLoop(ctx)
+		if err != nil {
+			level.Error(om.logger).Log("msg", "failed to initialize runtime configuration notify loop", "err", err)
+			return err
+		}
+	} else {
+		return errors.Errorf("Unsupported runtime configuration change detector: %s", om.cfg.ChangeDetector)
+	}
+
+	return nil
+}
+
+// notifyLoop utilizes file notifications to reduce the amount of reads it performs
+// it will conditionally resort to poll reading when encountering errors loading the configuration
+func (om *Manager) notifyLoop(ctx context.Context) error {
+	// This is tricky to get right, see the inotify man page and https://github.com/fsnotify/fsnotify/issues/372 for a list of issues
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	// If the file is a symlink follow it, otherwise modifications could be missed
+	fp, err := filepath.EvalSymlinks(om.cfg.LoadPath)
+	if err != nil {
+		return err
+	}
+
+	// Watch the directory rather than the file directly to handle the file's remove/add, since inodes are watched not names
+	dir, _ := filepath.Split(fp)
+	err = watcher.Add(dir)
+	if err != nil {
+		return err
+	}
+
+	done := make(chan bool)
+
+	go func() {
+		defer func() {
+			done <- true
+		}()
+
+		// It's okay to stop polling if the file doesn't exist since another event will happen upon creation
+		stopPredicate := func(err error) bool {
+			return err == nil || errors.Is(err, fs.ErrNotExist)
+		}
+
+		// Attempt to load the config first since no changes may arrive
+		om.pollLoop(ctx, stopPredicate)
```

Comment:
if I understand this correctly then this stops looping once the file has been loaded successfully, because the `err` passed into `stopPredicate` will be `nil`, right? 
I think an explanatory comment would be useful, because `pollLoop()` sounds like this is going to keep looping...

Reply:
Your understanding is correct. The nice thing about the original read loop was that transient errors didn't really matter because retries were going to happen anyway. The notification loop can't ignore an error and wait for another notification that may never arrive, so it uses the pool loop until it can read the file or see it doesn't exist.

I'll comment this more clearly. I could see the `stopPredicate` being interpreted as for the notification loop rather than the read loop at first glance as well. I should probably rename `pollLoop` to `readLoop` if I keep referring to it as read loop as well.

Reply:
I changed the name of `pollLoop` to `periodicReload`, `stopPredicate` to `stopReloadingPredicate` and changed the comments on this line to:
```go
	// Load the config or see it doesn't exist since no future changes may arrive
	// This call will return when either of the above holds because of stopReloadingPredicate
	om.periodicReload(ctx, stopReloadingPredicate)

```
I also separated out the internal notify loop from the setup, so hopefully the flow is easier to follow.